### PR TITLE
Fix #7336: MyFaces DataTable#filterAndSort

### LINKED ISSUE #7336: DataTable: Removed data (rows) remain after update when filtered before
**Describe the defect**
When using a List (for example ArrayList) of DTOs for a dataTable and removing some elements from the list, updating the table throw ajax, everythink works fine.
But when I first filter the rows inside the dataTable throw column or global Filter and then remove some of the still visible elements, they remain after ajax update. But the elements are arleady removed from the Java List.

The Bug was also discussed in the forum: https://forum.primefaces.org/viewtopic.php?f=3&t=66918

**Reproducer**
Checkout https://github.com/MatthiasPischka/primefaces-test.git and run as described (mvn jetty:run for example).
After starting, visit the page http://localhost:8080/primefaces-test/data-table-remove-bug.xhtml (you get there throw navigation from default page, too).
Working and buggy example are described on the page.

**Environment:**
 - PF Version: _10.0.0_
 - JSF + version: _Mojarra 2.3.14.SP01_
 - Affected browsers: _Testet in Firefox and Chrome_

### PR DESCRIPTION
**Pull request by christophs78**

### CODE REVIEW

**Comment by melloware** [comment]
So are we expecting this failure until its fixed?
```
Error:  Failures: 
Error:    DataTable021Test.testGlobalFilterAndRemoveRowV2:237 expected: <1> but was: <2>
```

**Comment by christophs78** [comment]
Yes. `DataTable#filterAndSort` does not work on MyFaces.
Maybe @tandraschko has some idea. Or we need to remove `DataTable#filterAndSort` we just added a few months ago.

**Comment by tandraschko** [comment]
can you explain what exactly does not work?

**Comment by christophs78** [comment]
I hoped it would be clear from looking at the failing integrationtest...
I try to describe.

DataTable is bound to a list auf 5 ProgrammingLanguages we remove one with the following code:
```java
public void removeV2(ProgrammingLanguage programmingLanguage) {
	this.progLanguages.remove(programmingLanguage);
	DataTable dataTable = (DataTable) FacesContext.getCurrentInstance().getViewRoot().findComponent("form:datatable");
	dataTable.filterAndSort();
}
```

When we debug FilterFeature#filter (called by DataTable#filterAndSort) it looks on Mojarra like this:
Mojarra_FilterFeature

On MyFaces (2.3; NOT next) it looks like this:
MyFaces_FilterFeature

This results in MyFaces calculating filteredValue-List wrong.

**Comment by tandraschko** [comment]
it only fails in 2.3 but not in 2.3-next?

**Comment by christophs78** [comment]
Can‘t say for 2.3-next. (Because wie can not replace MyFaces within TomEE yet.)

**Comment by christophs78** [comment]
Look´s like `table.setValue(null);` has "limitations" on MyFaces. (setValue is implemented by UIData)

Mojarra is (in this point) straight forward:
```java
public void setValue(Object value) {
	setDataModel(null);
	getStateHelper().put(PropertyKeys.value, value);
}
```

MyFaces 2.3 tries to do some magic which seems to be worse in this usecase:
```java
public void setValue(Object value)
{
	getStateHelper().put(PropertyKeys.value, value );
	_dataModelMap.clear();
	_rowStates.clear();
	_isValidChilds = true;
}
```

**Comment by tandraschko** [comment]
further research would be great

File: primefaces/src/main/java/org/primefaces/component/datatable/DataTable.java [hunk 1]
```diff
@@ -1100,10 +1100,17 @@ public void setWidth(String width) {
...
+        if (isLazy() || getFilteredValue() == null) {
             return;
         }
 
+        /*
+         * setDataModel is defined by UIData. So different implementations for Mojarra and MyFaces.
+         * But PrimeFaces comes with it´s own UIData which extends/modifies UIData provided by JSF-impl.
+         * But PrimeFaces UIData does not know all impl-specifics, so ....
+         */
+        setDataModel(null); // for MyFaces 2.3 - compatibility
```
Comment (christophs78) [inline primefaces/src/main/java/org/primefaces/component/datatable/DataTable.java:1112]: Seems like we have kind of a solution.
But hard to find good words for PrimeFaces-UIData-thing.

**Comment by tandraschko** [comment]
is the PR ready now?

**Comment by christophs78** [comment]
> is the PR ready now?

Yes

**Comment by melloware** [comment]
Looks like this is intermittently failing on MyFACES 2.3.X?  
```
 DataTable021Test.testGlobalFilterAndRemoveRowV1:278 » StaleElementReference st...
```

**Comment by christophs78** [comment]
Think i have some timing-issue. (does not happy always & happens for different tests)

```
Error:  Errors: 
Error:    DataTable021Test.testGlobalFilterAndRemoveRowV1:278 » StaleElementReference st...

Error:  Errors: 
Error:    DataTable021Test.testGlobalFilterAndRemoveRowV2:335 » StaleElementReference st...
```

**Pull request merged**